### PR TITLE
Visual updates, fix skip mode (NaN mode 1)

### PR DESCRIPTION
**Pull request by Chasyxx**
Skip mode used to ~~halt the renderer at the NaN handling stage~~ *works fine, but I changed it from polynomial to linear time*. Now if skip mode is used, NaN samples are actually skipped.

Also adds a progress bar, worker status indicators, and an "auto" option for the worker count.

image

### CODE REVIEW

File: sample_config.ini [hunk 1]
```diff
@@ -23,7 +23,7 @@ bits=8
 [Advanced]
 ; alternative length constant (based on the t value itself), seconds is ignored if not 0
 tLength=0
-; the amount of workers that will be working simultaneously (to speed up the process), set to 1 for expressions with a reverb function (or those that dont use t)
+; the amount of workers that will be working simultaneously (to speed up the process), set to 1 for expressions with a reverb function (or those that dont use a fake t). can be "auto"
```
Comment (Chasyxx) [inline sample_config.ini:26]: Accidential negative, "(...that ~~dont~~ use a fake t)."

**Comment by cabfile** [comment]
i dont see the point of "auto" amount of workers, since each worker is supposed to represent one CPU core, and your code will create up to 10 workers
you could've added a "max" setting instead, that would for example use 16 workers if the user's CPU has 16 cores
also, in the worker script, i wanted to use as little if/switch statements as possible (because i thought it would make the script run slower), unlike what you are doing
although i will say, the progress bar is nice, but maybe replace the active workers list with just a count of them?

**Comment by cabfile** [comment]
oh and, the invalidSamples mode 1 "fix" isnt even a fix, its just a different way of doing it

**Comment by Chasyxx** [comment]
> You could've added a "max" setting instead, that would for example use 16 workers if the user's CPU has 16 cores

I like the idea, but I don't understand how you'd find out how many cores are usable

> Also, in the worker script, I wanted to use as little if/switch statements as possible (because I thought it would make the script run slower), unlike what you are doing

and
> oh and, the invalidSamples mode 1 "fix" isn't even a fix, it's just a different way of doing it

Well, with my [test music](https://chasyxx.github.io/EnBeat_NEW/#EnBeat2-5VRBbsMgEPwLBws7CFgMiamUHPuAXJMcsOtDJbuOYg6Vovy9C07VRGlLe7ZWswszixAjsWfSDC8teSJ+7QVwWBm2f6Pva0p9oTgsix0w4HLJJLdxcdj5zQZMVh6ilApu/pVVEiXTD9UgFIbGXej6qX5/V4r7U07FZFuZgVpF4yT7HbOIW1NyoefgS+p9EXe+MDq+4o8UlTBc2hJ9WkwEGG5tob/opj556oI4NWgu1VXNg+6aYaQBKH3yQuWL2F6BVQI/Pc6AKoNwwOSEkdH1x67dOo8zQmuQkpF+GhjP3eB83TpPLh8), it seemed to be faster. The old method never really "skipped" NaNs, it just removed them afterward, and if there's an unusual amount that can look like a freeze to someone. By putting it into the worker process itself, might seem to make the worker slower and probably does a little bit, but for my test song the processing took 4-sh seconds, and then it said "looking for invalid samples" and didn't seem to do anything. It's an issue with 
1. looping through the whole array, and 
2. if a NaN is found, loop through the whole array and remove it. 
In the worst case, it has ~~O(n^2)~~ **O(n log n)**. By implementing it into the already O(n) worker, it remains O(n) and might be a bit slower. If anything, i might say mine might be better, but mine really only speeds things up with a large number if NaNs and maybe otherwise it would be better to stick with the old version (if a single if statement somehow takes any actual amount of time as to build up in normal music). I might do a couple of tests to see if that is the case

> although I will say, the progress bar is nice, but maybe replace the active workers list with just a count of them?

I can do that.

**Comment by Chasyxx** [comment]
2.031s
10.559s (probably my laptop doing something else rq)
2.520s
1.940s
1.897s

With if checks:
5.343s (probably another outlier)
2.424s
1.977s
2.110s
1.748s

rendering 32k floatbeat track, 2 workers, 15 seconds.
The if statement and try...catch have little or negligible impact here, but i understand your concern and will try to keep if and switch layers down

**Pull request closed without merging**